### ISSUE TITLE
gradient of the LP problem

### ISSUE THREAD

**Issue by SILI1994** [icoz69/DeepEMD#6]
Hi! Thank you for the nice work. I have a question about the gradient when using the QPTH solver to solve the LP problem. In my test, the gradients are almost always zeros. I attached the following testing code in the `models/emd_utils.py` file.

```python
if __name__ == '__main__':

    batch_size = 50
    num_node = 25

    cosine_distance_matrix = torch.rand(batch_size, num_node, num_node, device='cuda', requires_grad=True)
    weight1 = torch.rand(batch_size, num_node, device='cuda', requires_grad=True)
    weight2 = torch.rand(batch_size, num_node, device='cuda', requires_grad=True)

    _, qpth_flow = emd_inference_qpth(cosine_distance_matrix, weight1, weight2)
    qpth_flow.sum().backward()
    print(weight1.grad)
    print(weight2.grad)
    print(cosine_distance_matrix.grad)
```

The output gradient w.r.t all three terms are very small regardless of whether `L2` or `QP` form is used. So I am wondering how is the back-propagation done in such a condition. Would you like to tell me if there is a misunderstanding of me?

**Comment by icoz69** [comment]
hi, thank you for your interests in our work.
it is a good observation from your demo code with a toy example, but for our task the generated flow must be used together with the similarity matrix to generate the final score, which is further used for cross-entropy loss. if you directly sum the flow, the grad back-propagated does seem very small in the example.
 I just do a test where I add a detach() to the similarity maxtrix in https://github.com/icoz69/DeepEMD/blob/a8e4ef2c65ab9d95e9db4f466671dc537915b5ca/Models/models/Network.py#L128 , and then do the training. I can still observe grad with reasonable magnitude generated in the layers after loss is backward(), which means the qpth solver does generate useful gradients. 
Moreover, the improvement of the qpth solver over open-cv solver also indicates that informative gradients are generated by qpth.

**Comment by icoz69** [comment]
you may also consider it from another view that, if you sum the flow as the loss, you are actually encouraging a small overall flow. however, the total flow is a fixed value, so the gradients generated from it is meaningless thus closed to zero.

**Comment by SILI1994** [comment]
Thank you for your reply! I now get the point.
